File: _practice/2192-All-Ancestors-of-a-Node-in-a-Directed-Acyclic-Graph/dfs-reverse/dfs-reverse.js
// sam with a little bit of editorial

/**
 * @param {number} n
 * @param {number[][]} edges
 * @return {number[][]}
 */
var getAncestors = function (n, edges) {
  const res = Array(n)
    .fill(null)
    .map(() => new Set());

  const reversed_graph = {};
  for (const [s, d] of edges) {
    if (!reversed_graph[d]) reversed_graph[d] = [];
    reversed_graph[d].push(s);
  }

  const visited = new Set();

  const dfs = (node) => {
    if (node in reversed_graph) {
      for (const nextNode of reversed_graph[node]) {
        if (!visited.has(nextNode)) {
          dfs(nextNode);
          visited.add(nextNode);
        }
      }
    }
  };

  for (let i = 0; i < n; i++) {
    if (i in reversed_graph) {
      dfs(i);
      res[i] = Array.from(visited).sort((a, b) => a - b);
      visited.clear();
    } else {
      res[i] = [];
    }
  }

  return res;
};

File: _practice/2192-All-Ancestors-of-a-Node-in-a-Directed-Acyclic-Graph/dfs-reverse/dfs-reverse.py
from typing import List
from collections import defaultdict

class Solution:
    def getAncestors(self, n: int, edges: List[List[int]]) -> List[List[int]]:
        res = [[] for i in range(n)]
        graph = defaultdict(list)
        for s, d in edges:
            graph[d].append(s)

        visited = set()

        def dfs(node):
            for nextNode in graph[node]:
                if nextNode not in visited:
                    dfs(nextNode)
                    visited.add(nextNode)

        for i in range(n):
            if i in graph:
                dfs(i)
                for j in range(n):
                    if j in visited:
                        res[i].append(j)
                visited.clear()        

        return res